File: 1051 Pop Sequence.cpp
#include<iostream>
#include<stack>
#include<queue>
using namespace std;

int n, k;
unsigned int m;

bool check(queue<int> q){
    int i;
    stack<int> in_s;
    stack<int> s;
    
    for (i=n; i>=1; i--){
        in_s.push(i);
    }
    
    while (!q.empty()){
        if ((!s.empty()) && (s.top() == q.front())){
            s.pop();
            q.pop();
        }
        else if (!in_s.empty()){
            s.push(in_s.top());
            in_s.pop();
            if (s.size() > m){
                return false;
            }
        }
        else if (in_s.empty()){
            return false;
        }
    }
    return true;
}

int main(){
    int i, j;
    cin>>m>>n>>k;
    
    for (i=0; i<k; i++){
        queue<int> q;
        for (j=0; j<n; j++){
            int x;
            cin>>x;
            q.push(x);
        }
        if (check(q)){
            cout<<"YES"<<endl;
        }
        else {
            cout<<"NO"<<endl;
        }
    }
    
    return 0;
}

File: 1051 Pop Sequence.py
M, N, K = map(int, input().split())

def check(q):
    in_s = list(range(N, 0, -1))
    s = []
    
    while q:
        if s and s[-1] == q[-1]:
            s.pop()
            q.pop()
        elif in_s:
            s.append(in_s.pop())
            if len(s) > M:
                return False
        elif not in_s:
            return False
    return True

if __name__ == "__main__":
    for _ in range(K):
        q = list(map(int, input().split()))
        q.reverse()
        if check(q):
            print("YES")
        else:
            print("NO")